### PR TITLE
Feature plugins

### PR DESCRIPTION
**Pull request by herbepau**
@dangershony @bokobza 

Jeremy / Dan - please look at this AppsFeature PR.

* AppsFeature - the new feature
* AppsFileService - responsible for any file I/O e.g. reading from the Stratis apps folder
* AppsHost - responsible for hosting and hosting lifetime of the single page app
* AppsStore - responsible for storing the apps as read from config (injected into AppsController)
* StratisApp - represents an app as defined in app config
* AppsController - ASP controller injected with IAppStore

- Tests are at Stratis.Bitcoin.Features.Apps.Tests
- Stratis.Bitcoin.Apps.Browser - trial Blazor app that communicates with the apps feature via the AppsController

As discussed with Jeremy, have moved the FindPorts method from a test file to new IpHelper class.  This is now used by the aforementioned test and the new apps feature.

Thanks
-Paul

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Api/Startup.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ public void ConfigureServices(IServiceCollection services)
 
                             builder
                             .WithOrigins(allowedDomains)
+                            .AllowAnyOrigin()
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Api/Startup.cs]: yes this should be changed!

[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Api/Startup.cs]: Yes, sorry, this will be removed - was to see the Blazor app browser working.  We don't need this

File: src/Stratis.StratisD/Program.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using Stratis.Bitcoin.Builder;
 using Stratis.Bitcoin.Configuration;
 using Stratis.Bitcoin.Features.Api;
+using Stratis.Bitcoin.Features.Apps;
```
[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.StratisD/Program.cs]: So the new app feature has UseApps - this namespace reference was left in here and can be removed.

File: src/Stratis.StratisD/Stratis.StratisD.csproj [hunk 3]
```diff
@@ -22,6 +22,7 @@
   <ItemGroup>
     <ProjectReference Include="..\NBitcoin\NBitcoin.csproj" />
     <ProjectReference Include="..\Stratis.Bitcoin.Features.Api\Stratis.Bitcoin.Features.Api.csproj" />
+    <ProjectReference Include="..\Stratis.Bitcoin.Features.Apps\Stratis.Bitcoin.Features.Apps.csproj" />
```
[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.StratisD/Stratis.StratisD.csproj]: Same as previous/above

File: src/Stratis.StratisD/Program.cs [hunk 4]
```diff
@@ -28,9 +29,9 @@ public class Program
                     .UseBlockStore()
                     .UsePosConsensus()
                     .UseMempool()
-                    .UseWallet()
+                    .UseWallet()                    
                     .AddPowPosMining()
-                    .UseApi()
+                    .UseApi()
```
[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.StratisD/Program.cs]: This is simply to do with the adding/removing of UseApps during development.

File: src/Stratis.Bitcoin.Features.Api/Startup.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Api/Startup.cs]: Have removed.

File: src/Stratis.Bitcoin.IntegrationTests.Common/EnvironmentMockUpHelpers/CoreNode.cs [hunk 5]
```diff
@@ -59,11 +59,6 @@ public CoreNode(NodeRunner runner, NodeBuilder builder, string configfile)
             this.ConfigParameters.Import(builder.ConfigParameters);
             this.ports = new int[3];
             IpHelper.FindPorts(this.ports);
-
-            var loggerFactory = new ExtendedLoggerFactory();
-            loggerFactory.AddConsoleWithFilters();
-
-            this.networkPeerFactory = new NetworkPeerFactory(network, DateTimeProvider.Default, loggerFactory, new PayloadProvider().DiscoverPayloads(), new SelfEndpointTracker());
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.IntegrationTests.Common/EnvironmentMockUpHelpers/CoreNode.cs]: Is this change related to this PR?

Comment (bokobza) [reply src/Stratis.Bitcoin.IntegrationTests.Common/EnvironmentMockUpHelpers/CoreNode.cs]: I think it's a merge thing that was left behind.

File: src/Stratis.Bitcoin.Features.Api/Startup.cs [hunk 6]
```diff
@@ -40,7 +40,7 @@ public void ConfigureServices(IServiceCollection services)
                             var allowedDomains = new[] { "http://localhost", "http://localhost:4200" };
 
                             builder
-                            .WithOrigins(allowedDomains)
+                            .WithOrigins(allowedDomains)
```
[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Api/Startup.cs]: Done.  + removed that Browser project.  Have pushed for further review.  Thanks

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                this.logger.LogWarning(
+                    $"No Stratis applications found at or below {this.appsFileService.StratisAppsFolderPath}");
+                return;
+            }
+
+            this.applications.AddRange(apps);
+        }
+
+        private IStratisApp CreateApp(FileInfo fileInfo)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: `CreateApp` sounds like you're creating an app rather than loading one.
Would `LoadApp` be more appropriate?

File: src/Stratis.Bitcoin.Features.Apps/AppsFileService.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.Configuration;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.Features.Apps.Interfaces;
+
+namespace Stratis.Bitcoin.Features.Apps
+{
+    public class AppsFileService : IAppsFileService
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Apps/AppsFileService.cs]: You can probably do without the AppsFileService.

We have a class, `Stratis.Bitcoin.Utilities.FileStorage<T>` where you can do 
```
this.fileStorage = new FileStorage<AppConfig>(dataFolder.ApplicationsPath);
IEnumerable<AppConfig> appsConfigs = this.fileStorage.LoadByFileExtension(".json"); // or ".app.json" if you fancy
```

Have a look at how we do it in wallets [here](https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs#L31), [here](https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs#L121) and [here](https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs#L157).

Also, I'm keen for the AppConfig object to be a simple model with attributes that you can just load and use, like in 
```
[JsonProperty(PropertyName = "displayName")]
public string DisplayName { get; set; }

[JsonProperty(PropertyName = "webRoot")]
public string WebRoot{ get; set; }
```
rather than get the fields like so:

```
var displayName = this.appsFileService.GetConfigSetting(fileInfo, "displayName");
var webRoot = this.appsFileService.GetConfigSetting(fileInfo, "webRoot");
```

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        public IEnumerable<IStratisApp> Applications => this.applications;
+
+        private void Load()
+        {
+            IEnumerable<FileInfo> configs = this.appsFileService.GetStratisAppConfigFileInfos();
+            var apps = configs.Select(this.CreateApp).Where(x => x != null).ToList();
+
+            if (apps.IsEmpty())
+            {
+                this.logger.LogWarning(
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: As a matter of policy we don't use string interpolation in logs (for performance).
You can do `this.logger.LogWarning("No Stratis applications found at or below {0}", "this.appsFileService.StratisAppsFolderPath");`.

File: src/Stratis.Bitcoin.Apps.Browser/Dto/StratisApp.cs [hunk 10]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Stratis.Bitcoin.Apps.Browser.Dto
+{
+    public class StratisApp
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Apps.Browser/Dto/StratisApp.cs]: Please have a look at my comment in AppsFileService.cs.

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs (hunk 7 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: It's a private method that creates an instance of a StratisApp per app in config, ready for hosting.  No loading involved.

File: src/Stratis.Bitcoin.Features.Apps/AppsFileService.cs (hunk 8 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsFileService.cs]: OK, I'll take a look

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs (hunk 9 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: OK

File: src/Stratis.Bitcoin.Apps.Browser/Dto/StratisApp.cs (hunk 10 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Apps.Browser/Dto/StratisApp.cs]: How does this Blazor project class relate to the comment ?

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 11]
```diff
@@ -28,38 +29,37 @@ public AppsStore(ILoggerFactory loggerFactory, IAppsFileService appsFileService,
...
+                if (this.applications.IsEmpty())
+                    this.logger.LogWarning("No Stratis applications found at or below {0}",
+                        this.dataFolder.ApplicationsPath);
+            }
+            catch (Exception e)
             {
-                this.logger.LogWarning(
-                    $"No Stratis applications found at or below {this.appsFileService.StratisAppsFolderPath}");
-                return;
+                this.logger.LogError($"Failed to load Stratis apps : {0}", e.Message);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: Don't use interpolated strings in logs - https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/Documentation/using-logging.md#interpolated-strings

Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: can remove $

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 12]
```diff
@@ -28,38 +29,37 @@ public AppsStore(ILoggerFactory loggerFactory, IAppsFileService appsFileService,
...
-                    stratisApp.WebRoot = webRoot;
-
+                StratisApp stratisApp = fileStorage.LoadByFileName(ConfigFileName);
+                stratisApp.Location = fileStorage.FolderPath;
                 return stratisApp;
             }
             catch (Exception e)
             {
-                this.logger.LogError($"Failed to create app : {e.Message}");   
+                this.logger.LogError($"Failed to create app : {0}", e.Message);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: can remove $, also why space before `:`?

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs (hunk 11 above)
Comment (Aprogiena) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: also why space before `:`?

File: src/Stratis.StratisD/Program.cs [hunk 13]
```diff
@@ -30,7 +31,8 @@ public class Program
                     .UseMempool()
                     .UseWallet()
                     .AddPowPosMining()
-                    .UseApi()
+                    .UseApi()        
+                    .UseApps()
```
Comment (Aprogiena) [inline src/Stratis.StratisD/Program.cs]: I don't think we want to use apps in default Stratis Demon

File: src/Stratis.StratisD/Stratis.StratisD.csproj (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aprogiena) [reply src/Stratis.StratisD/Stratis.StratisD.csproj]: But I can still see it

File: src/Stratis.StratisD/Program.cs (hunk 13 above)
Comment (bokobza) [reply src/Stratis.StratisD/Program.cs]: I think we do want to have Apps as a first class feature.
The idea is that soon we won't be looking at the console anymore but rather at a web page, so I don't mind this coming in now.

Comment (dangershony) [reply src/Stratis.StratisD/Program.cs]: Yeah I agree I would have it as part of the default deamon (same as api)

Comment (bokobza) [reply src/Stratis.StratisD/Program.cs]: Happy for it to be left out until the sec issues are resolved.

Comment (bokobza) [reply src/Stratis.StratisD/Program.cs]: Then it'll be in by default 💃

Comment (Aprogiena) [reply src/Stratis.StratisD/Program.cs]: It is the principle rather than current issues. I mean, it won't be ever solved, attack surface will always be greater as well as resource reqs.

File: src/Stratis.Bitcoin.Apps.Browser/Interfaces/IAppsService.cs [hunk 14]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Stratis.Bitcoin.Apps.Browser.Dto;
+
+namespace Stratis.Bitcoin.Apps.Browser.Interfaces
+{
+    public interface IAppsService
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Apps.Browser/Interfaces/IAppsService.cs]: Hi Paul please add comments on this interface to describe the intention

File: src/Stratis.Bitcoin.Apps.Browser/Services/AppsService.cs [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+        public AppsService(HttpClient httpClient)
+        {
+            this.httpClient = httpClient;
+        }
+
+        public async Task<List<StratisApp>> GetApplicationsAsync()
+        {
+            var serializer = new DataContractJsonSerializer(typeof(List<StratisApp>));
+            Task<System.IO.Stream> streamTask = this.httpClient.GetStreamAsync("http://localhost:38221/api/apps/all");
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Apps.Browser/Services/AppsService.cs]: We need to think on how we inject the port at startup.

File: src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs [hunk 16]
```diff
@@ -0,0 +1,65 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Builder;
+using Stratis.Bitcoin.Builder.Feature;
+using Stratis.Bitcoin.Configuration.Logging;
+using Stratis.Bitcoin.Features.Apps.Interfaces;
+
+namespace Stratis.Bitcoin.Features.Apps
+{
+    public class AppsFeature : FullNodeFeature
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs]: Please comment this and describe the intent of what we are trying todo and achieve

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+        private readonly List<IStratisApp> applications = new List<IStratisApp>();
+        private readonly DataFolder dataFolder;
+        private const string ConfigFileName = "stratisApp.json";
+
+        public AppsStore(ILoggerFactory loggerFactory, DataFolder dataFolder)
+        {
+            this.dataFolder = dataFolder;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.Load();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: Not in the constructor, DI constructor should do bear minimal

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public IEnumerable<IStratisApp> Applications => this.applications;
+
+        private void Load()
+        {
+            try
+            {
+                FileInfo[] fileInfos = new DirectoryInfo(this.dataFolder.ApplicationsPath)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: We probably want to look in a specific folder no?

File: src/Stratis.Bitcoin.Features.Apps/AppsHost.cs [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        private readonly List<(IStratisApp app, IWebHost host)> hostedApps = new List<(IStratisApp app, IWebHost host)>();
+
+        public AppsHost(ILoggerFactory loggerFactory) 
+        {
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);            
+        }
+
+        public IEnumerable<IStratisApp> HostedApps => this.hostedApps.Select(x => x.app);        
+
+        public void Host(IEnumerable<IStratisApp> stratisApps) =>
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Apps/AppsHost.cs]: Too long for a one liner lambda style.

File: src/Stratis.Bitcoin/Configuration/DataFolder.cs [hunk 20]
```diff
@@ -27,6 +27,7 @@ public DataFolder(string path)
             this.RpcCookieFile = Path.Combine(path, ".cookie");
             this.WalletPath = Path.Combine(path);
             this.LogPath = Path.Combine(path, "Logs");
+            this.ApplicationsPath = Path.Combine(path, "apps");
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Configuration/DataFolder.cs]: Ah ok ignore my last comment on the path

File: src/Stratis.Bitcoin/Utilities/IpHelper.cs [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    public static class IpHelper
+    {
+        /// <summary>
+        /// Find ports that are free to use.
+        /// </summary>
+        /// <param name="ports">A list of ports to checked or fill/replace as necessary.</param>
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Utilities/IpHelper.cs]: Not clear comment, is this an existing list of ports?

**Comment by bokobza** [comment]
In the Stratis.Bitcoin.Browser.Apps, can you please remove all the things that are not related to our code, like the survey and the weather stuff?

File: src/Stratis.StratisD/Program.cs (hunk 13 above)
Comment (dangershony) [reply src/Stratis.StratisD/Program.cs]: But this is a light wait project and nothing is added by default yet (infect I think we should load apps only on demand)
This will allow us to provide a set of default UIs to interact with the node without having to plug new apps.

Regarding security that's another topic, yeah the security issue  comes from the ability to plug new apps so if one does not plug any then there is not added risk.

File: src/Stratis.Bitcoin/Utilities/IpHelper.cs (hunk 21 above)
Comment (bokobza) [reply src/Stratis.Bitcoin/Utilities/IpHelper.cs]: You pass this method a list of ports, and this method will either:
- replace the ports that are used by new, unused ones
- add new, unused ports to the list

File: src/Stratis.StratisD/Program.cs (hunk 13 above)
Comment (Aprogiena) [reply src/Stratis.StratisD/Program.cs]: I'm not totally against it, like I would block it or something, just want to make sure everyone is on the same page about consequences of having it. I'd personally not include it, but if you are OK with all the consequences, then OK.

Comment (Aprogiena) [reply src/Stratis.StratisD/Program.cs]: What I mean is that I wouldn't include apps into the node at all, I would rather go with UI being separated application instead of browser. But it looks like this is too much conceptually different than what you are looking for.

File: src/Stratis.Bitcoin/Utilities/IpHelper.cs (hunk 21 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Utilities/IpHelper.cs]: ok yeah, I was just referring to the comment not being clear on the behaviour.

File: src/Stratis.StratisD/Program.cs (hunk 13 above)
Comment (dangershony) [reply src/Stratis.StratisD/Program.cs]: This could anyway be achieved if the separate UI would be in another processes using RPC to communicate with the node (which is similar to browser I guess).

Agree we need to carefully inspect what apps are delivered with the node itself. 
This is not different to the current swagger UI that is part of the node already.

Comment (Aprogiena) [reply src/Stratis.StratisD/Program.cs]: yeah, kind of hate swagger inside node :D

File: src/Stratis.Bitcoin.Apps.Browser/Interfaces/IAppsService.cs (hunk 14 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Apps.Browser/Interfaces/IAppsService.cs]: OK

File: src/Stratis.Bitcoin.Apps.Browser/Services/AppsService.cs (hunk 15 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Apps.Browser/Services/AppsService.cs]: Indeed

File: src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs (hunk 16 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsFeature.cs]: OK

File: src/Stratis.Bitcoin.Features.Apps/AppsHost.cs [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Features.Apps.Interfaces;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Apps
+{
+    public class AppsHost : IDisposable, IAppsHost
+    {        
+        private readonly ILogger logger;
+        private readonly List<(IStratisApp app, IWebHost host)> hostedApps = new List<(IStratisApp app, IWebHost host)>();
```
[the remark this answers is not in the document]
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsHost.cs]: You mean new the list in the ctor ?  OK

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs (hunk 18 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: Nope - this code is simply looking for all stratisApp.json files under the apps folder

File: src/Stratis.Bitcoin.Features.Apps/AppsHost.cs (hunk 19 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsHost.cs]: OK

File: src/Stratis.Bitcoin/Configuration/DataFolder.cs (hunk 20 above)
Comment (herbepau) [reply src/Stratis.Bitcoin/Configuration/DataFolder.cs]: Yep

File: src/Stratis.Bitcoin/Utilities/IpHelper.cs (hunk 21 above)
Comment (herbepau) [reply src/Stratis.Bitcoin/Utilities/IpHelper.cs]: I simply took the method and comment from TestsHelper, but yeah a beefier comment maybe

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs (hunk 11 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: done

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs (hunk 12 above)
Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: done

File: src/Stratis.Bitcoin.Features.Apps/AppsStore.cs [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+                return this.applications;
+            }
+        }
+
+        private void Load()
+        {
+            try
+            {
+                Debug.Assert(this.applications==null);
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: Instead of throwing an exception here, can you return the applications instead?

Comment (herbepau) [reply src/Stratis.Bitcoin.Features.Apps/AppsStore.cs]: Not quite, but have made cleaner. 

Just pushed.

**Comment by dangershony** [comment]
Please fix conflicts

**Comment by bokobza** [comment]
Looks good to me.
@herbepau I created a local change to fix these merge conflicts.
If you want I can push these to this PR.
